README.md
# petit_httpd

[![pub package](https://img.shields.io/pub/v/petit_httpd.svg?logo=dart&logoColor=00b9fc)](https://pub.dev/packages/petit_httpd)
[![Null Safety](https://img.shields.io/badge/null-safety-brightgreen)](https://dart.dev/null-safety)
[![Codecov](https://img.shields.io/codecov/c/github/gmpassos/petit_httpd)](https://app.codecov.io/gh/gmpassos/petit_httpd)
[![Dart CI](https://github.com/gmpassos/petit_httpd/actions/workflows/dart.yml/badge.svg?branch=master)](https://github.com/gmpassos/petit_httpd/actions/workflows/dart.yml)
[![GitHub Tag](https://img.shields.io/github/v/tag/gmpassos/petit_httpd?logo=git&logoColor=white)](https://github.com/gmpassos/petit_httpd/releases)
[![New Commits](https://img.shields.io/github/commits-since/gmpassos/petit_httpd/latest?logo=git&logoColor=white)](https://github.com/gmpassos/petit_httpd/network)
[![Last Commits](https://img.shields.io/github/last-commit/gmpassos/petit_httpd?logo=git&logoColor=white)](https://github.com/gmpassos/petit_httpd/commits/master)
[![Pull Requests](https://img.shields.io/github/issues-pr/gmpassos/petit_httpd?logo=github&logoColor=white)](https://github.com/gmpassos/petit_httpd/pulls)
[![Code size](https://img.shields.io/github/languages/code-size/gmpassos/petit_httpd?logo=github&logoColor=white)](https://github.com/gmpassos/petit_httpd)
[![License](https://img.shields.io/github/license/gmpassos/petit_httpd?logo=open-source-initiative&logoColor=green)](https://github.com/gmpassos/petit_httpd/blob/master/LICENSE)

This is a simple HTTP file server integrated with [Let's Encrypt][shelf_letsencrypt], [gzip][shelf_gzip] and CORS.

[shelf_gzip]: https://pub.dev/packages/shelf_gzip
[shelf_letsencrypt]: https://pub.dev/packages/shelf_letsencrypt

## Motivation

Since [Dart][dart_overview_platforms] can run in many native platforms (Linux/x64, macOS/x64/arm64, Windows/x86),
it can be an awesome way to have an HTTP Server running in any place,
including all the basic features needed today.

[dart_overview_platforms]: https://dart.dev/overview#platform

## API Documentation

See the [API Documentation][api_doc] for a full list of functions, classes and extension.

[api_doc]: https://pub.dev/documentation/petit_httpd/latest/

## Usage

```dart
import 'dart:io';

import 'package:petit_httpd/petit_httpd.dart';

void main() async {
  var petitHTTPD = PetitHTTPD(Directory('/var/www'),
      port: 8080,
      securePort: 443,
      bindingAddress: '0.0.0.0',
      letsEncryptDirectory: Directory('/etc/letsencrypt/live'),
      domains: {'mydomain.com': 'contact@mydomain.com'});

  var ok = await petitHTTPD.start();

  if (!ok) {
    print('** ERROR Starting: $petitHTTPD');
    exit(1);
  }

  print('-- STARTED: $petitHTTPD');
}
```

## CLI Tool

### petit_httpd

The `petit_httpd` is a **CLI** for the [PetitHTTPD class][PetitHTTPD_class].

[PetitHTTPD_class]: https://pub.dev/documentation/petit_httpd/latest/petit_httpd/PetitHTTPD-class.html

First activate the `petit_httpd` command:
```shell
$> dart pub global activate petit_httpd
```

To run an HTTP Daemon:

```shell
$> petit_httpd ./www --port 8080 --securePort 443 --address 0.0.0.0 --letsencrypt-path /etc/letsencrypt/live --domain mydomain.com
```

A BASH script with all together:
```shell
#!/bin/bash

dart pub global activate petit_httpd

export PATH="$PATH":"$HOME/.pub-cache/bin"

mkdir -p /var/log/petit_httpd

petit_httpd /var/www --address 0.0.0.0 --letsencrypt-path /etc/letsencrypt/live --domain mydomain.com -verbose  >> /var/log/petit_httpd/requests.log
```

## Source

The official source code is [hosted @ GitHub][github_petit_httpd]:

- https://github.com/gmpassos/petit_httpd

[github_petit_httpd]: https://github.com/gmpassos/petit_httpd

# Features and bugs

Please file feature requests and bugs at the [issue tracker][tracker].

# Contribution

Any help from the open-source community is always welcome and needed:
- Found an issue?
    - Please fill a bug report with details.
- Wish a feature?
    - Open a feature request with use cases.
- Are you using and liking the project?
    - Promote the project: create an article, do a post or make a donation.
- Are you a developer?
    - Fix a bug and send a pull request.
    - Implement a new feature, like other training algorithms and activation functions.
    - Improve the Unit Tests.
- Have you already helped in any way?
    - **Many thanks from me, the contributors and everybody that uses this project!**

*If you donate 1 hour of your time, you can contribute a lot,
because others will do the same, just be part and start with your 1 hour.*

[tracker]: https://github.com/gmpassos/petit_httpd/issues

# Author

Graciliano M. Passos: [gmpassos@GitHub][github].

[github]: https://github.com/gmpassos

## License

[Apache License - Version 2.0][apache_license]

[apache_license]: https://www.apache.org/licenses/LICENSE-2.0.txt


CHANGELOG.md
## 1.1.0

- sdk: '>=3.0.0 <4.0.0'
- shelf_letsencrypt: ^1.2.0
- lints: ^2.1.0

## 1.0.6

- Fix call to `LetsEncrypt.startSecureServer`.
- shelf: ^1.4.1
- shelf_static: ^1.1.2
- shelf_gzip: ^4.0.1
- shelf_letsencrypt: ^1.1.1
- lints: ^2.0.1
- test: ^1.24.1
- dependency_validator: ^3.2.2
- coverage: ^1.6.3
- pubspec: ^2.3.0
- path: ^1.8.3
- mercury_client: ^2.1.8

## 1.0.5

- Added `-cache-control` parameter.
- shelf: ^1.4.0
- shelf_static: ^1.1.1
- sdk: '>=2.17.0 <3.0.0' 

## 1.0.4

- mercury_client: ^2.1.5

## 1.0.3

- `petit_httpd` new option:
  - `-force-https`: to automatically redirect to HTTPS when possible.
  - `-version`: just show the current CLI version. 

## 1.0.2

- Added server name in headers.

## 1.0.1

- Added `CORS` and `cache-control` headers.

## 1.0.0

- Initial version.


test/petit_httpd_test.dart
import 'dart:io';

import 'package:mercury_client/mercury_client.dart';
import 'package:path/path.dart' as pack_path;
import 'package:petit_httpd/petit_httpd.dart';
import 'package:test/test.dart';

void main() {
  group('PetitHTTPD', () {
    test('basic', () async {
      var documentRoot = _resolveDocumentRoot();

      expect(documentRoot.existsSync(), isTrue);

      var petitHTTPD = PetitHTTPD(documentRoot, port: 4455);

      var ok = await petitHTTPD.start();
      expect(ok, isTrue);

      var httpClient = HttpClient('http://localhost:4455/');

      {
        var response = await httpClient.get('/foo');
        print(response);

        expect(response.isNotOK, isTrue);
        expect(response.status, equals(404));
        expect(response.getResponseHeader('cache-control'), isNull);

        expect(response.getResponseHeader('server'),
            equals('petit_httpd/${PetitHTTPD.VERSION}'));
      }

      {
        var response = await httpClient.get('/hello-world.txt');
        print(response);

        expect(response.isOK, isTrue);
        expect(response.status, equals(200));
        expect(response.getResponseHeader('cache-control'),
            contains('must-revalidate'));

        expect(response.bodyAsString, equals('Hello World!'));

        expect(response.getResponseHeader('server'),
            equals('petit_httpd/${PetitHTTPD.VERSION}'));
      }
    });
  });
}

Directory _resolveDocumentRoot() {
  for (var p in ['./', './test', '../test']) {
    var file = File(pack_path.join(p, 'doc-root/hello-world.txt'));
    if (file.existsSync()) {
      return file.parent.absolute;
    }
  }

  return Directory('.');
}


test/petit_httpd_version_test.dart
@TestOn('vm')
import 'dart:io';

import 'package:path/path.dart' as path;
import 'package:pubspec/pubspec.dart';
import 'package:test/test.dart';

void main() {
  group('PetitHTTPD.VERSION', () {
    setUp(() {});

    test('Check Version', () async {
      var projectDirectory = Directory.current;

      print(projectDirectory);

      var pubspecFile = File(path.join(projectDirectory.path, 'pubspec.yaml'));

      print('pubspecFile: $pubspecFile');

      var pubSpec = await PubSpec.loadFile(pubspecFile.path);

      print('PubSpec.name: ${pubSpec.name}');
      print('PubSpec.version: ${pubSpec.version}');

      var srcPath =
          path.join(projectDirectory.path, 'lib/src/petit_httpd_base.dart');

      var srcFile = File(srcPath);

      print(srcFile);

      var src = srcFile.readAsStringSync();

      var versionMatch = RegExp(r"VERSION\s*=\s*'(.*?)'").firstMatch(src)!;

      var srcVersion = versionMatch.group(1);

      print('srcVersion: $srcVersion');

      expect(pubSpec.version.toString(), equals(srcVersion),
          reason:
              'Bones_API.VERSION[$srcVersion] != PubSpec.version[${pubSpec.version}]');
    });
  });
}


lib/src/petit_httpd_base.dart
import 'dart:io';

import 'package:shelf/shelf.dart';
import 'package:shelf/shelf_io.dart' as shelf_io;
import 'package:shelf_gzip/shelf_gzip.dart';
import 'package:shelf_letsencrypt/shelf_letsencrypt.dart';
import 'package:shelf_static/shelf_static.dart';

/// A Simple HTTP Daemon.
class PetitHTTPD {
  // ignore: constant_identifier_names
  static const String VERSION = '1.1.0';

  /// The document root [Directory].
  final Directory documentRoot;

  /// The port number. Default: `80`.
  final int port;

  /// The secure port (HTTPS) number. Default: `443`.
  final int securePort;

  /// The server socket binding address: Default: `localhost`.
  final String bindingAddress;

  /// If `true` the `Cache-Control` header will be set for successful responses.
  final bool setHeaderCacheControl;

  /// Personalized value for the header `Cache-Control`.
  /// Will be used if [setHeaderCacheControl] is `true`.
  final String? headerCacheControl;

  /// If `true` CORS` headers will be set for successful responses.
  final bool setCORSHeaders;

  /// If [true] and a [SecureServer] is being started
  ///
  /// - See [isLetsEncryptEnabled].
  final bool redirectToHTTPS;

  /// If `true` allows Let's Encrypt.
  final bool allowLetsEncrypt;

  /// The domains to use with Let's Encrypt.
  final Map<String, String>? domains;

  /// The Let's Encrypt [Directory]
  final Directory? letsEncryptDirectory;

  PetitHTTPD(this.documentRoot,
      {this.port = 80,
      this.securePort = 443,
      this.bindingAddress = 'localhost',
      this.setHeaderCacheControl = true,
      this.headerCacheControl,
      this.setCORSHeaders = true,
      this.redirectToHTTPS = true,
      this.allowLetsEncrypt = true,
      this.domains,
      this.letsEncryptDirectory});

  /// Returns `true` if Let's Encrypt is enabled and all prerequisites are accomplished.
  bool get isLetsEncryptEnabled =>
      allowLetsEncrypt &&
      letsEncryptDirectory != null &&
      domains != null &&
      domains!.isNotEmpty;

  /// Starts the daemon.
  Future<bool> start() async {
    if (isLetsEncryptEnabled) {
      return _startLetsEncrypt();
    } else {
      return _startNormal();
    }
  }

  Future<bool> _startNormal() async {
    var handler = _createShelfHandler();

    var server = await shelf_io.serve(handler, bindingAddress, port);

    _configureServer(server);

    return true;
  }

  void _configureServer(HttpServer server) {
    // Enable built-in [HttpServer] gzip:
    server.autoCompress = true;
  }

  Handler _createShelfHandler({bool withSecureServer = false}) {
    var pipeline = const Pipeline().addMiddleware(logRequests());

    if (withSecureServer && redirectToHTTPS) {
      pipeline = pipeline.addMiddleware(_redirectToHttpsMiddleware);
    }

    pipeline = pipeline
        .addMiddleware(gzipMiddleware)
        .addMiddleware(_headersMiddleware);

    var handler = pipeline.addHandler(
        createStaticHandler(documentRoot.path, defaultDocument: 'index.html'));

    return handler;
  }

  Handler _redirectToHttpsMiddleware(Handler innerHandler) {
    return (request) {
      var requestedUri = request.requestedUri;

      if (requestedUri.scheme == 'http') {
        final domains = this.domains;
        if (domains != null && domains.containsKey(requestedUri.host)) {
          var secureUri = requestedUri.replace(scheme: 'https');
          return Response.seeOther(secureUri);
        }
      }

      return innerHandler(request);
    };
  }

  Handler _headersMiddleware(Handler innerHandler) {
    return (request) {
      return Future.sync(() => innerHandler(request))
          .then((response) => _configureHeaders(request, response));
    };
  }

  Response _configureHeaders(Request request, Response response,
      {Map<String, String>? headers}) {
    headers ??= <String, String>{};

    headers[HttpHeaders.serverHeader] = 'petit_httpd/$VERSION';

    var statusCode = response.statusCode;
    if (statusCode < 200 || statusCode > 299) {
      return response.change(headers: headers);
    }

    if (setHeaderCacheControl) {
      var headerCacheControl = this.headerCacheControl?.trim();

      var cacheControl = headerCacheControl != null &&
              headerCacheControl.isNotEmpty
          ? headerCacheControl
          : 'private, must-revalidate, max-age=1, stale-while-revalidate=60';

      headers[HttpHeaders.cacheControlHeader] = cacheControl;
    }

    if (setCORSHeaders) {
      headers = setHeadersCORS(request, response, headers: headers);
    }

    return response.change(headers: headers);
  }

  static const String headerXAccessToken = "X-Access-Token";
  static const String headerXAccessTokenExpiration =
      "X-Access-Token-Expiration";
  static const String headerXAccessTokenSource = "X-Access-Token-Source";

  static const String headerAccessControlRequestHeaders =
      "Access-Control-Request-Headers";
  static const String headerAccessControlRequestMethod =
      "Access-Control-Request-Method";

  static const String exposeHeaders =
      "Content-Length, Content-Type, Last-Modified, $headerXAccessToken, $headerXAccessTokenExpiration";

  static Map<String, String> setHeadersCORS(Request request, Response response,
      {Map<String, String>? headers}) {
    headers ??= <String, String>{};

    var origin = getOrigin(request);

    var localhost = false;

    if (origin == null || origin.isEmpty) {
      headers["Access-Control-Allow-Origin"] = "*";
    } else {
      headers["Access-Control-Allow-Origin"] = origin;

      if (origin.contains("://localhost:") ||
          origin.contains("://127.0.0.1:") ||
          origin.contains("://::1")) {
        localhost = true;
      }
    }

    headers["Access-Control-Allow-Methods"] =
        "GET,HEAD,PUT,POST,PATCH,DELETE,OPTIONS";
    headers["Access-Control-Allow-Credentials"] = "true";

    if (localhost) {
      headers["Access-Control-Allow-Headers"] =
          "Content-Type, Access-Control-Allow-Headers, Authorization, x-ijt";
    } else {
      headers["Access-Control-Allow-Headers"] =
          "Content-Type, Access-Control-Allow-Headers, Authorization";
    }

    headers["Access-Control-Expose-Headers"] = exposeHeaders;

    return headers;
  }

  static String? getRemoteAddress(Request request) {
    final connectionInfo =
        request.context['shelf.io.connection_info'] as HttpConnectionInfo?;
    return connectionInfo?.remoteAddress.address;
  }

  static String? getOrigin(Request request) {
    var origin = request.headers['origin'];
    if (origin != null) return origin;

    var requestedUri = request.requestedUri;

    var host = request.headers['host'];
    if (host != null) {
      origin = "${requestedUri.scheme}://$host/";
      return origin;
    }

    var remoteHost = getRemoteAddress(request) ?? host;
    if (remoteHost == null) return null;

    origin = "${requestedUri.scheme}://$remoteHost/";
    return origin;
  }

  Future<bool> _startLetsEncrypt() async {
    final certificatesHandler = CertificatesHandlerIO(letsEncryptDirectory!);

    final LetsEncrypt letsEncrypt =
        LetsEncrypt(certificatesHandler, production: true);

    var handler = _createShelfHandler(withSecureServer: true);

    final domains = this.domains;

    var domain = domains!.keys.first;
    var domainEmail = domains[domain] ?? 'contact@$domain';

    var servers = await letsEncrypt.startSecureServer(
      handler,
      {domain: domainEmail},
      port: port,
      securePort: securePort,
      bindingAddress: bindingAddress,
    );

    var server = servers[0]; // HTTP Server.
    var serverSecure = servers[1]; // HTTPS Server.

    _configureServer(server);
    _configureServer(serverSecure);

    return true;
  }

  @override
  String toString() {
    var letsEncryptInfo = isLetsEncryptEnabled
        ? ', redirectToHTTPS: $redirectToHTTPS, domains: $domains, letsEncryptDirectory: $letsEncryptDirectory'
        : ', letsEncrypt: disabled';
    return 'PetitHTTPD{ documentRoot: $documentRoot, port: $port, securePort: $securePort, bindingAddress: $bindingAddress$letsEncryptInfo }';
  }
}
